packages/next/src/server/web/spec-extension/unstable-cache.ts

```diff
@@ -158,7 +158,9 @@ export function unstable_cache<T extends Callback>(
 
         if (cacheEntry) {
           const resData = cacheEntry.value.data
-          cachedValue = JSON.parse(resData.body)
+          // JSON.stringify(undefined) returns `undefined`, but JSON.parse(undefined) throws an error
+          cachedValue =
+            resData.body !== undefined ? JSON.parse(resData.body) : undefined
         }
```

Comment:
Can we also just move this entire condition below `if (isStale) { ... }` ?

Reply:
makes sense, updated

Reply:
The function was refactored, but the issue of [`JSON.stringify`ing the cache's data](https://github.com/vercel/next.js/blob/c754d9fb6ecddc7cf46a61548dc8c49539c398d1/packages/next/src/server/web/spec-extension/unstable-cache.ts#L35) that could be `undefined`, then [`JSON.parse`ing it](https://github.com/vercel/next.js/blob/c754d9fb6ecddc7cf46a61548dc8c49539c398d1/packages/next/src/server/web/spec-extension/unstable-cache.ts#L186) on retrieval persists, so I'm going to leave this PR open.

Reply:
I'm not sure why is this not being merged or reviewed yet by NextJs team 🤔
Isn't this a critical bug, when an api returns `undefined` the `unstable_cache` crashes the Application. Isn't that a very common scenario??

cc: @mattddean , @ijjk , @timneutkens